### ISSUE TITLE
Resolve: Diffusers版とBFL版のLoRAの変換スクリプト書いてたけど...

### ISSUE THREAD

**Issue by kaeru-shigure** [kohya-ss/sd-scripts#1467]
> [Kohya Tech](https://x.com/kohya_tech) [@kohya_tech](https://x.com/kohya_tech) · [5時間](https://x.com/kohya_tech/status/1824804530472751216)
Diffusers版とBFL版のLoRAの変換スクリプト書いてたけど、よく考えたら qkv = nn.Linear(dim, dim * 3) へのLoRAと、to_q = nn.Linear(dim, dim) , to_v, to_k へのLoRA 3個、変換できないんじゃね(;･∀･)

@kohya-ss (*ﾟーﾟ)

```py
import torch
import torch.nn as nn
n = 5
m = 7
to_qkv = nn.Linear(n, m * 3)
to_q = nn.Linear(n, m)
to_k = nn.Linear(n, m)
to_v = nn.Linear(n, m)
input = torch.randn(11, n)

# to_qkv -> to_q, to_k, to_v
for i, (weight, bias) in enumerate(zip(
    to_qkv.weight.split(m),
    to_qkv.bias.split(m),
)):
    (to_q, to_k, to_v)[i].state_dict()["weight"].copy_(weight)
    (to_q, to_k, to_v)[i].state_dict()["bias"].copy_(bias)

## to_q, to_k, to_v -> to_qkv
#to_qkv.state_dict()["weight"].copy_(torch.cat([
#    to_q.weight,
#    to_k.weight,
#    to_v.weight,
#], dim=0))
#to_qkv.state_dict()["bias"].copy_(torch.cat([
#    to_q.bias,
#    to_k.bias,
#    to_v.bias,
#], dim=0))

assert torch.allclose(
    to_qkv(input),
    torch.cat([
        to_q(input),
        to_k(input),
        to_v(input),
    ], dim=-1)
  )
```

**Comment by kaeru-shigure** [comment]
確かに。lora_down/lora_upにおいてはshapeじゃなくてsizeが変わってしまってるので互換性無いですね。q/k/v->qkvの場合、dimを上げるしかない。問題の本質を理解しました。
仰っている解決策しかなさそうです。LoRAのデータサイズは増加しますが互換性は保てますね。
拡大変換時に使わないweightを0fillしておけばis_converted_from_separated_qkv = weight[:,src_dim:].all(0)的な処理で判定できるので可逆変換として戻すことはできそうです。
